design/client_conformance_profiles.md

```diff
@@ -0,0 +1,550 @@
+
+When registering the newly client or configuring the client that has already been registered, the client conformance profile checks whether the configuration of the client complies with the applied security profile.
+
+If not, keycloak rejects this registration/configuration and returns the appropriate error.
+
+* Client's Request Check
+
+When the client sends the request, the client conformance profile checks whether this request complies with the applied security profile.
+
+If not, keycloak rejects this request and returns the appropriate error.
+
+* Security Profile Specific Operation on the Client's Request
+
+Depending on the security profile, keycloak needs to do this profile specific operation when it receives the request from the client.
+
+## Scope
+
+### Target Security Profile
+
+The following security profiles are to be supported.
+
+* [FAPI (Financial-grade API) Read Only API Security Profile](
+https://openid.net/specs/openid-financial-api-part-1-ID2.html)
+* [FAPI (Financial-grade API) Read and Write API Security Profile](https://openid.net/specs/openid-financial-api-part-2-ID2.html)
+
+The following FAPI based security profiles might be supported in the future.
+
+* UK : [OpenBanking Security Profile](
+https://openbanking.atlassian.net/wiki/spaces/DZ/pages/7046134/Open+Banking+Security+Profile+-+Implementer+s+Draft+v1.1.0)
+* Australia : [CDR (Consumer Data Rights) Security Profile](
+https://consumerdatastandardsaustralia.github.io/standards/#security-profile)
+
+
+## Design
+
+### Scope of client conformance profile
+
+Keycloak can select and set the client conformance profile per client.
+
+### The number of applicable client conformance profile simultaneously
+
+Keycloak can set only one client conformance profile to the client.
+
+### When and how keycloak selects and sets the client conformance profile to the client
+
+The administrator must be involved in the process of determining and setting the client conformance profile to the client.
+
+It must be prohibited for the client itself to determine and set the client conformance profile by itself, without the administrator's involvement.
+
+The ways of client registration are as follows.
+
+* Dynamic Client Registration Endpoint
+    - Keycloak Client Representation
+    - OpenID Connect Client Metadata Description
+
+* Client Registration CLI
+
+* Admin Console
+    - UI
+    - CLI
+    - REST API directly
+    - Import
+ 
+The following 2 ways of client registrations can only determine and set the client conformance profile to the client.
+
+* Dynamic Client Registration Endpoint
+    - OpenID Connect Client Metadata Description
+        * Use Initial Access Token
+
+When creating the initial access token on the admin console, keycloak provides the pull-down menu for specifying the client conformance profile. The selected client conformance profile is included onto the initial access token to convey it to dynamic client registration endpoint afterwards.

```

Comment:
I'm not sure whether clients that are enforced different FAPI security profile coexist in the same realm. However, IMO, it might be probable to do so. 

It might be probable to create the realm per bank and accommodate the client that only access its Account Service API and the one that also access its Payment Service API. In that case, the former be enforced FAPI Read Only Security Profile while the latter be enforced FAPI Read and Write Security Profile.

Meanwhile, considering the point that complete Client Conformance Profiles implementation needs a lot of code change and effort, it seems to be reasonable to implement this Client Conformance Profiles step by step.

Therefore, how about implementing client registration check per realm by using Client Registration Policy as you said at first and implementing client registration check per client later as follow-up task?

BTW, this design document states that implementing client registration check can be conducted on the following two points.

* Dynamic Client Registration Endpoint
    - OpenID Connect Client Metadata Description
        * Use Initial Access Token

* Admin Console
    - UI

Client Registration Policy covers the former. As for the latter, We need to do the same check as Client Registration Policy. Therefore, how about implementing client registration check on the Dynamic Client Registration Endpoint at first as follows

* Dynamic Client Registration Endpoint
    - OpenID Connect Client Metadata Description
        * Use Initial Access Token
        * Anonymous

 and implementing client registration check on Admin Console later as follow-up task?